### PR TITLE
semi auto with parameters

### PR DESCRIPTION
**Pull request by a2sandres**
* Modificacion del modo de vuelo Auto
* Desarrollado por Aerospace Scannig Technologies AERSOCANTECH.
 * Programado por el ingeniero de sistemas ANDRES ARDILA SEDANO apodo a2sAndres.
 * Esta Modificacion del modo de vuelo   Auto se basa en el modo de vuelo Alt_hold, la cual  permite el control manual de altitud mientra se realiza un plan de vuelo.
 * Al entrar en el control de altura la aeronave se detiene
 * reinicia el avance hacia adelante dándole entrada de tono y continuando con el plan de vuelo.
 * Se activa esta funcion con el parametro Auto Manual Altitude Control AUTO_MAN_ALT donde sus valores son 0:Disabled,1:Enabled.
* Permite tambien el bloqueo de curso (YAW) point in the direction next waypoint (no pilot input accepted), con el parametro Enabled Course Locked ENBL_CRS_LOCK donde sus valores son 0:Disabled,1:Enabled.

Esta es una opcion, para que se agreguen estas funciones sin agregar un modo de vuelo adicional

### CODE REVIEW

**Comment by khancyr** [comment]
duplicate of https://github.com/ArduPilot/ardupilot/pull/9577 ? This need to be in english ...

**Comment by a2sandres** [comment]
* Modification of the Auto flight mode
* Developed by Aerospace Scannig Technologies AERSOCANTECH.
* Programmed by systems engineer ANDRES ARDILA SEDANO nickname a2sAndres.
* This Modification of the Auto flight mode is based on the Alt_hold flight mode, which allows manual altitude control while a flight plan is made.
* When entering the height control the aircraft stops
* Restart the forward advance giving tone entry and continuing with the flight plan.
* This function is activated with the parameter Auto Manual Altitude Control AUTO_MAN_ALT where its values ​​are 0: Disabled, 1: Enabled.
* It also allows course blocking (YAW) point in the direction next waypoint (no pilot input accepted), with the Enabled Course Locked parameter ENBL_CRS_LOCK where its values ​​are 0: Disabled, 1: Enabled.

This is an option, so that these functions are added without adding an additional flight mode

**Comment by a2sandres** [comment]
although the same behavior is not a copy of #9577, since the behavior is activated with parameters and not with the individual flight mode, in addition to this parameter the behavior is added: course enabled blocked.
This behavior was created with the intention of being applied in drone spraying.

**Comment by Pedals2Paddles** [comment]
You need to rebase on master and add your changes. I'm not sure what you did here, because it shows all the changes as brand new files with 4,000+ new lines of code. It's impossible to actually see what changed.

**Comment by Pedals2Paddles** [comment]
That said, I like the idea and concept.

**Comment by a2sandres** [comment]
thank you I will review it

**Comment by a2sandres** [comment]
Suggested corrections have already been made to review the changes.

**Comment by a2sandres** [comment]
hi I would like to review the conflicts presented by the compiler and that these are part of the new preogramcion

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -214,6 +214,20 @@ const AP_Param::Info Copter::var_info[] = {
     // @User: Standard
     GSCALAR(wp_yaw_behavior,  "WP_YAW_BEHAVIOR",    WP_YAW_BEHAVIOR_DEFAULT),
 
+    // @Param: AUTO_MAN_ALT
+    // @DisplayName: Auto Manual Altitude Control
+    // @Description: Flight mode developed by Aerospace Scannig Technologies AERSOCANTECH. Programmed by Systems Engineer ANDRES ARDILA SEDANO nick name a2sAndres.This flight mode is based on the Auto and Alt_hold flight modes, which allows manual altitude control.While entering the height control the aircraft stops.it re starts the forward advance by giving it pitch input and continuing with the flight plan
```
Comment (OXINARF) [inline ArduCopter/Parameters.cpp]: This initial part of the description needs to be removed on both parameters. Also:

- needs spaces after a period
- this isn't a new flight mode, so description isn't accurate

File: ArduCopter/Parameters.h [hunk 2]
```diff
@@ -370,8 +370,12 @@ class Parameters {
 
         // 254,255: reserved
 
+        // Semi auto mode param
+        k_param_auto_man_alt = 256, // activate semi auto mode add a2sAndres
+        k_param_enbl_crs_lock, // activate enabled course locked func add a2sAndres
```
Comment (OXINARF) [inline ArduCopter/Parameters.h]: Can't use this space, you need to use G2. Also remove the `add a2sAndres`.

File: ArduCopter/mode_auto.cpp [hunk 3]
```diff
@@ -56,7 +56,7 @@ bool Copter::ModeAuto::init(bool ignore_checks)
 //      relies on run_autopilot being called at 10hz which handles decision making and non-navigation related commands
 void Copter::ModeAuto::run()
 {
-    // call the correct auto controller
+	// call the correct auto controller
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong change

File: ArduCopter/Parameters.h [hunk 4]
```diff
@@ -370,8 +370,12 @@ class Parameters {
...
+
         // the k_param_* space is 9-bits in size
-        // 511: reserved
+        // 511: reserved
```
Comment (OXINARF) [inline ArduCopter/Parameters.h]: Please don't add whitespace.

File: ArduCopter/mode_auto.cpp [hunk 5]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
-    if (!copter.failsafe.radio) {
-        // get pilot's desired yaw rate
-        target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
-        if (!is_zero(target_yaw_rate)) {
-            auto_yaw.set_mode(AUTO_YAW_HOLD);
-        }
-    }
 
-    // set motors to full range
+    //Enable course locked add a2sandres
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Please add space after `//` and remove `add a2sAndres`.

File: ArduCopter/Parameters.h [hunk 6]
```diff
@@ -432,6 +436,8 @@ class Parameters {
     AP_Int8         flight_mode6;
     AP_Int8         simple_modes;
     AP_Int8         flight_mode_chan;
+    AP_Int32        auto_man_alt;
```
Comment (OXINARF) [inline ArduCopter/Parameters.h]: Why is this an AP_Int32?

File: ArduCopter/config.h [hunk 7]
```diff
@@ -384,6 +384,19 @@
...
+//
+#ifndef AUTO_MAN_ALT
+# define AUTO_MAN_ALT   ENABLED
+#endif
+
+//////////////////////////////////////////////////////////////////////////////
+// ENBL_CRS_LOCK
+//
+#ifndef ENBL_CRS_LOCK
+# define ENBL_CRS_LOCK   ENABLED
```
Comment (OXINARF) [inline ArduCopter/config.h]: Both features should be disabled by default.

File: ArduCopter/mode_auto.cpp [hunk 8]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+    if (g.enbl_crs_lock == 1) {
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: The code for this feature doesn't make sense to me at all. The code says that if the feature is enabled and there **is** pilot yaw input then we look at next waypoint. The description of the parameter doesn't say that though.

File: ArduCopter/mode_auto.cpp [hunk 9]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+        if (!copter.failsafe.radio) {
+            // get pilot's desired yaw rate
+            target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+            if (!is_zero(target_yaw_rate)) {
+                auto_yaw.set_mode(AUTO_YAW_LOOK_AT_NEXT_WP);
+            }
+        }
+    }
+    else {
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: else needs to be on previous line.

File: ArduCopter/mode_auto.cpp [hunk 10]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+    }
+
+   // set motors to full range
     motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // run waypoint controller 
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());  
+
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong changes above, whitespaces were added.

File: ArduCopter/mode_auto.cpp [hunk 11]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+   // set motors to full range
     motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // run waypoint controller 
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());  
+      
+     // mode semi auto
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Line has one extra space at beginning.

File: ArduCopter/mode_auto.cpp [hunk 12]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+            target_climb_rate = get_surface_tracking_climb_rate(target_climb_rate, pos_control->get_alt_target(), G_Dt);
+        }
 
+        // update altitude target and call position controller Alt_Hold add a2sAndres
+        pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
+    }
+        
     // call z-axis position controller (wpnav should have already updated it's alt target)
-    pos_control->update_z_controller();
+    pos_control->update_z_controller();
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong change, whitespaces added at the end.

File: ArduCopter/mode_auto.cpp [hunk 13]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
 
     // call attitude controller
-    if (auto_yaw.mode() == AUTO_YAW_HOLD) {
-        // roll & pitch from waypoint controller, yaw rate from pilot
-        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), target_yaw_rate);
-    } else {
-        // roll, pitch from waypoint controller, yaw heading from auto_heading()
-        attitude_control->input_euler_angle_roll_pitch_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), auto_yaw.yaw(),true);
+    //Enable course locked add a2sandres
+    if (g.enbl_crs_lock == 1) {
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Why is this needed?

File: ArduCopter/mode_auto.cpp [hunk 14]
```diff
@@ -821,7 +887,7 @@ void Copter::ModeAuto::spline_run()
         // get pilot's desired yaw rat
         target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
         if (!is_zero(target_yaw_rate)) {
-            auto_yaw.set_mode(AUTO_YAW_HOLD);
+            auto_yaw.set_mode(AUTO_YAW_HOLD);
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong change.

File: ArduCopter/mode_auto.cpp [hunk 15]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // run waypoint controller 
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());  
+      
+     // mode semi auto
+    if (g.auto_man_alt == 1) {
+        
+        // get pilot desired climb rate alt control Alt_Hold add a2sAndres
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Remove all `add a2sAndres` from comments.

File: ArduCopter/mode_auto.cpp [hunk 16]
```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
...
+        float target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+        target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
+
+        // get pilot desired lean angles Althold add a2sAndres
+        float target_roll, target_pitch;
+        get_pilot_desired_lean_angles(target_roll, target_pitch, copter.aparm.angle_max, attitude_control->get_althold_lean_angle_max());
+
+        //added control to resume flight add a2sAndres
+        if (target_roll != 0 || target_pitch != 0) {
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: What is this needed for? We never changed modes...

File: ArduCopter/mode_auto.cpp [hunk 17]
```diff
@@ -1072,13 +1138,19 @@ void Copter::ModeAuto::do_nav_wp(const AP_Mission::Mission_Command& cmd)
         }
     }
 
+    if(target_loc.alt!=current_loc.alt){
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Needs spaces:

- after if
- before and after `!=`
- after `)`

File: ArduCopter/mode_auto.cpp [hunk 18]
```diff
@@ -1072,13 +1138,19 @@ void Copter::ModeAuto::do_nav_wp(const AP_Mission::Mission_Command& cmd)
...
+    
     // this will be used to remember the time in millis after we reach or pass the WP.
     loiter_time = 0;
     // this is the delay, stored in seconds
     loiter_time_max = cmd.p1;
 
     // Set wp navigation target
-    wp_start(target_loc);
+    wp_start(target_loc);   
+
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong change, whitespaces were added.

File: ArduCopter/mode_auto.cpp [hunk 19]
```diff
@@ -1285,6 +1357,7 @@ void Copter::ModeAuto::do_guided_limits(const AP_Mission::Mission_Command& cmd)
         cmd.content.guided_limits.alt_min * 100.0f,    // convert meters to cm
         cmd.content.guided_limits.alt_max * 100.0f,    // convert meters to cm
         cmd.content.guided_limits.horiz_max * 100.0f); // convert meters to cm
+
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: Wrong change, unneeded new line added.

File: ArduCopter/mode_auto.cpp [hunk 20]
```diff
@@ -1072,13 +1138,19 @@ void Copter::ModeAuto::do_nav_wp(const AP_Mission::Mission_Command& cmd)
         }
     }
 
+    if(target_loc.alt!=current_loc.alt){
+        target_loc.alt = current_loc.alt;
```
Comment (OXINARF) [inline ArduCopter/mode_auto.cpp]: This isn't mentioned in the parameter description, it changes the altitude of the current waypoint target to be whatever altitude the pilot is controlling it to get to.
This doesn't work with splines (or other navigation commands).

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (a2sandres) [reply ArduCopter/Parameters.cpp]: Revised and changed

File: ArduCopter/config.h (hunk 7 above)
Comment (a2sandres) [reply ArduCopter/config.h]: revised and changed

File: ArduCopter/Parameters.h (hunk 4 above)
Comment (a2sandres) [reply ArduCopter/Parameters.h]: line 378 revised and changed

File: ArduCopter/Parameters.h (hunk 2 above)
Comment (a2sandres) [reply ArduCopter/Parameters.h]: revised and changed, use g2 added in the line 601,line. To add the parameters I use, use the wiki tutorial http://ardupilot.org/dev/docs/code-overview-adding-a-new-parameter.html, for which I suggest that this section be updated to be clarified when you should use the k_pararms and the G2 param, as well as add the g2 params

File: ArduCopter/mode_auto.cpp (hunk 3 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 5 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/Parameters.h (hunk 6 above)
Comment (a2sandres) [reply ArduCopter/Parameters.h]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 8 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: The ENB_CRS_LOCK behavior is eliminated because this function is performed by the WP_YAW_BEHAVIOR parameter

File: ArduCopter/mode_auto.cpp (hunk 9 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 10 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 11 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 12 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 13 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: The ENB_CRS_LOCK behavior is eliminated because this function is performed by the WP_YAW_BEHAVIOR parameter

File: ArduCopter/mode_auto.cpp (hunk 14 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: The ENB_CRS_LOCK behavior is eliminated because this function is performed by the WP_YAW_BEHAVIOR parameter

File: ArduCopter/mode_auto.cpp (hunk 15 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 16 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: added control to resume the return plan, this may be due to loss of information, once a height change of more or less than 5 meters has been reached, whereby the aircraft stops its advance forward. To resume the flight plan, enter the pitch or roll.

If you know any solution or a better change for this situation, I would appreciate it if you would tell me. but I have to tell you that we have already done tests with fumigational drones and with frames F550 to check this behavior and for the moment it has worked with the expected results, however we continue doing more rigorous tests for other activities and to be able to implement this behavior with more trust

File: ArduCopter/mode_auto.cpp (hunk 17 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 18 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 19 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

File: ArduCopter/mode_auto.cpp (hunk 20 above)
Comment (a2sandres) [reply ArduCopter/mode_auto.cpp]: revised and changed

**Comment by a2sandres** [comment]
The suggestions made by you were taken care of and explained, I remain pendite to any novelty

**Comment by OXINARF** [comment]
@a2sandres You've changed all lines in Parameters.h and mode_auto.cpp... Likely spaces changes. Also, as I mentioned before, you need to squash all commits and rebase in master, this is getting messy.

**Comment by magicrub** [comment]
@a2sandres if you need help fixing this PR, please let us know and we can help

**Comment by peterbarker** [comment]
I'm going to close this as the offers to help out weren't responded to way back in 2018.  The code has moved on significantly since then, and the up-coming s-curve navigation code will make whatever remains in here not mergeable.  Further, Randy has mentioned the possibility of doing "nudging" after s-curves is in, and that might take the place of the feature mooted here.

Thanks anyway!
